src/JsonAsynchronousNodeKit/LargeJsonParser.cs

```diff
@@ -0,0 +1,281 @@
+        }
+
+        this.stream.Position = GetStartPosition(this.stream);
+
+        if (!stream.CanRead || stream.Read(this.buffer) == 0)
+        {
+            throw new EndOfStreamException();
+        }
+
+        static int GetStartPosition(Stream stream)
+        {
+            var bom = Encoding.UTF8.Preamble.ToArray();
+            stream.Position = 0;
+            var buffer = new byte[bom.Length];
+            _ = stream.Read(buffer, 0, buffer.Length);
+
+            return Enumerable.SequenceEqual(buffer, bom) ? 3 : 0;
+        }
+    }
+
+    /// <summary>
+    /// Begin evaluating the next section of the json stream.
+    /// </summary>
+    /// <returns>A ParserStateResult object describing the section that was encountered or null if this was the final state.</returns>
+    /// <remarks>If the result object is an enumerable you MUST ensure that you've walked it entirely before calling next again.</remarks>
+    public ParserStateResult? Next()
+    {
+        if (this.enumeratorActive)
+        {
+            throw new ParserException("You must walk the entire enumerable from the previous result before calling Next() again.");
+        }
+
+        try
+        {
+            var reader = new Utf8JsonReader(this.buffer, isFinalBlock: this.isFinalBlock, this.readerState);
+
+            if (!this.isParsingStarted)
+            {
+                ParserUtils.SkipNoneTokens(this.stream, ref this.buffer, ref reader);
+
+                ParserUtils.AssertTokenType(this.stream, ref reader, JsonTokenType.StartObject);
+                ParserUtils.GetMoreBytesFromStream(this.stream, ref this.buffer, ref reader);
+
+                this.isParsingStarted = true;
+            }
+
+            ParserUtils.Read(this.stream, ref this.buffer, ref reader);
+
+            // If the end of the Json Object is reached, return null to indicate completion.
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                return null;
+            }
+            else if (this.previousResultState is not null && this.previousResultState.YieldReturn && reader.TokenType == JsonTokenType.EndArray)
+            {
+                // yield returning json arrays means we can't pass it the same Utf8JsonReader ref, so we need to create a new one.
+                // BUT when we do that we end up consuming the next token, so we need to leave it in the array case to be eatten by the next caller.
+                ParserUtils.Read(this.stream, ref this.buffer, ref reader);
+            }
+
+            // If the end of the Json Object is reached, return null to indicate completion.
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                return null;
+            }
+
+            ParserUtils.AssertTokenType(this.stream, ref reader, JsonTokenType.PropertyName);
+            var propertyName = reader.GetString() ?? throw new NotImplementedException();
```

Comment:
The null-coalescing is to coerse the type from string? to string (I don't approve of null-forgiveness if it can be avoided, though if that's the style in this repo I will abide by it), I'll change NotImplementedException to InvalidOperationException and give it a sensible title (that should never be seen.